File: ai/allocation/distance.py
"""Small deterministic geographic distance helper for allocation."""

from __future__ import annotations

import math

from ai.allocation.constraints import has_valid_coordinates


def calculate_distance_km(lat1: float | None, lon1: float | None, lat2: float | None, lon2: float | None) -> float | None:
    """Return Haversine distance in kilometers, or None when coordinates are unavailable."""
    if not all(has_valid_coordinates(latitude, longitude) for latitude, longitude in ((lat1, lon1), (lat2, lon2))):
        return None
    earth_radius_km = 6371.0088
    latitude_one, latitude_two = math.radians(lat1), math.radians(lat2)
    delta_latitude = math.radians(lat2 - lat1)
    delta_longitude = math.radians(lon2 - lon1)
    haversine = math.sin(delta_latitude / 2) ** 2 + math.cos(latitude_one) * math.cos(latitude_two) * math.sin(delta_longitude / 2) ** 2
    return round(earth_radius_km * 2 * math.asin(math.sqrt(max(0.0, min(1.0, haversine)))), 4)

File: maps/location/distance.ts
import type { Coordinate } from "./types";

const EARTH_RADIUS_METERS = 6_371_000;

export function distanceInMeters(
  first: Coordinate,
  second: Coordinate,
): number {
  const toRadians = (degrees: number) => (degrees * Math.PI) / 180;

  const latitude1 = toRadians(first.latitude);
  const latitude2 = toRadians(second.latitude);

  const deltaLatitude = toRadians(second.latitude - first.latitude);
  const deltaLongitude = toRadians(second.longitude - first.longitude);

  const a =
    Math.sin(deltaLatitude / 2) ** 2 +
    Math.cos(latitude1) *
      Math.cos(latitude2) *
      Math.sin(deltaLongitude / 2) ** 2;

  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

  return EARTH_RADIUS_METERS * c;
}
